File: algorithms/code/leetcode/lc225_implement_stack_using_queues/lc225_implement_stack_using_queues.py
# https://leetcode.com/problems/implement-stack-using-queues/

import collections
import unittest
from typing import Deque


class MyStack:
    def __init__(self) -> None:
        self.__active: Deque[int] = collections.deque()
        self.__other: Deque[int] = collections.deque()

    def __swap(self) -> None:
        swap = self.__other
        self.__other = self.__active
        self.__active = swap

    # Push element x onto stack.
    def push(self, x: int) -> None:
        # 1 - active
        self.__other.append(x)
        while self.__active:
            self.__other.append(self.__active.popleft())
        self.__swap()

    # Removes the element on top of the stack.
    def pop(self) -> int:
        return self.__active.popleft()

    # Get the top element.
    def top(self) -> int:
        return self.__active[0]

    # Return whether the stack is empty.
    def empty(self) -> bool:
        return not bool(self.__active)


# It's also possible with just one stack


class TestCode(unittest.TestCase):
    def test_example(self) -> None:
        solution = MyStack()
        solution.push(5)
        solution.push(2)
        assert solution.top() == 2
        solution.pop()
        assert solution.top() == 5
        assert not solution.empty()
        solution.pop()
        assert solution.empty()

File: algorithms/code/leetcode/lc225_implement_stack_using_queues/lc225_implement_stack_using_queues.rs
// https://leetcode.com/problems/implement-stack-using-queues/
// #easy

use std::collections::VecDeque;
use std::mem;

#[derive(Default)]
pub struct Solution {
    active: VecDeque<i32>,
    other: VecDeque<i32>,
}

impl Solution {
    pub fn new() -> Self {
        Self {
            active: VecDeque::new(),
            other: VecDeque::new(),
        }
    }

    fn swap(&mut self) {
        mem::swap(&mut self.active, &mut self.other);
    }

    // Push element x onto stack.
    pub fn push(&mut self, x: i32) {
        self.other.push_back(x);
        while !self.active.is_empty() {
            self.other.push_back(self.active.pop_front().unwrap());
        }
        self.swap();
    }

    // Removes the element on top of the stack.
    pub fn pop(&mut self) {
        self.active.pop_front();
    }

    // Get the top element.
    pub fn top(&self) -> i32 {
        *self.active.front().unwrap()
    }

    // Return whether the stack is empty.
    pub fn empty(&self) -> bool {
        self.active.is_empty()
    }

    // It's also possible with just one stack
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        let mut solution = Solution::new();
        solution.push(5);
        solution.push(2);
        assert_eq!(solution.top(), 2);
        solution.pop();
        assert_eq!(solution.top(), 5);
        assert!(!solution.empty());
        solution.pop();
        assert!(solution.empty());
    }
}
